### PR TITLE
fix: Port already bound Error

### PR DESCRIPTION
**Pull request by varshith257**
fixes https://github.com/zio/zio-http/issues/2776
/claim https://github.com/zio/zio-http/issues/2776

Here's the working demonstration:
https://www.loom.com/share/8a39b9440e4341ceb1137f3ea8ab6f2d?sid=84c7ed14-8e7a-4aca-98cc-be593198e939

### CODE REVIEW

**Comment by varshith257** [comment]
@jdegoes I have a similar experience working on a port already bound issue with the Kubernetes organization. I have implemented the retry mechanism to allocate port as same what I have done in Kubernetes too :)

It's ready for your review

**Comment by varshith257** [comment]
@jdegoes Have a look at this PR

**Comment by varshith257** [comment]
@jdegoes Could you take a time to review this PR?

**Comment by varshith257** [comment]
@jdegoes A gentle reminder to have a look on it 😄

**Comment by varshith257** [comment]
@jdegoes A small gentle reminder to have a look on it 😄

**Comment by varshith257** [comment]
@jdegoes PTAL

File: build.sbt [hunk 1]
```diff
@@ -41,6 +41,7 @@ lazy val root =
       `zio-quickstart-sql`
     )
     .settings(
+      run / fork := true,
```
Comment (guizmaii) [inline build.sbt:44]: Why do we need to add this?

Comment (varshith257) [reply build.sbt:44]: @guizmaii 
The reason for adding `run / fork := true` in the build.sbt file is to handle the issue where the server process does not release the port (in this case, port 8080) after being interrupted by Ctrl+C.

When `run / fork := true` is set, the server is run in a separate JVM process, which allows it to be properly terminated when Ctrl+C is used. This ensures that the port is released, preventing the `Port already bound` error when we try to run the server again on the same port.

Comment (guizmaii) [reply build.sbt:44]: Answering my own question. See https://github.com/zio/zio-http/issues/2776

Comment (guizmaii) [reply build.sbt:44]: @varshith257 Are these project using the [sbt-revolver](https://github.com/spray/sbt-revolver) plugin?

Comment (varshith257) [reply build.sbt]: Added!

**Review by guizmaii — APPROVED** [review]

**Review by khajavi — CHANGES_REQUESTED** [review]
This PR doesn't resolve the issue mentioned! Please open the PR whenever it is resolved!

```
sbt:zio-quickstart-restful-webservice> run
[info] running dev.zio.quickstart.MainApp
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
timestamp=2024-08-16T15:43:51.288677Z level=ERROR thread=#zio-fiber-1 message="" cause="Exception in thread "zio-fiber-10,6,4" io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
        at dev.zio.quickstart.MainApp.run(MainApp.scala:16)"
[error] Exception in thread "zio-fiber-10,6,4" io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
[error]         at dev.zio.quickstart.MainApp.run(MainApp.scala:16)
[error]         at dev.zio.quickstart.MainApp.run(MainApp.scala:16)
[error] Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
[error] stack trace is suppressed; run last Compile / run for the full output
[error] (Compile / run) Exception in thread "zio-fiber-10,6,4" io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
[error]         at dev.zio.quickstart.MainApp.run(MainApp.scala:16)
[error] Total time: 3 s, completed Aug 16, 2024, 7:13:51 PM
sbt:zio-quickstart-restful-webservice> run
[info] running dev.zio.quickstart.MainApp
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
timestamp=2024-08-16T15:44:09.636552Z level=ERROR thread=#zio-fiber-1 message="" cause="Exception in thread "zio-fiber-10,6,4" io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
        at dev.zio.quickstart.MainApp.run(MainApp.scala:16)"
[error] Exception in thread "zio-fiber-10,6,4" io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
[error]         at dev.zio.quickstart.MainApp.run(MainApp.scala:16)
[error]         at dev.zio.quickstart.MainApp.run(MainApp.scala:16)
[error] Caused by: io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
[error] stack trace is suppressed; run last Compile / run for the full output
[error] (Compile / run) Exception in thread "zio-fiber-10,6,4" io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
[error]         at dev.zio.quickstart.MainApp.run(MainApp.scala:16)
[error] Total time: 3 s, completed Aug 16, 2024, 7:14:09 PM
sbt:zio-quickstart-restful-webservice>
```

**Comment by varshith257** [comment]
@khajavi Have you tried with updated cmds in  README.md? 

sbt-resolver plugin is used for this

**Comment by khajavi** [comment]
@varshith257 
It is preferable not to change the commands, as we need to update all the tutorials on zio.dev referenced to this quickstart.

**Comment by varshith257** [comment]
It is the library suggested by @guizmaii which works good for this Port bind error. Or else my previous approach of retry logic

**Comment by khajavi** [comment]
@varshith257 
We need consistent material across zio.dev and zio-quickstarts. Please see this page: https://zio.dev/guides/quickstarts/restful-webservice
The sbt-respolver is a good plugin but If you propose using it, I suggest creating PR to update documentation.

**Comment by varshith257** [comment]
Ok, I will add docs for it in a note for this problem. I also mention how to configure this plugin instead the updation of here

**Comment by khajavi** [comment]
To reduce the number of changes and make it consistent with other quick starts, I suggest resolving this issue without switching to alternative commands other than the `sbt run`.

**Comment by varshith257** [comment]
What about port retry logic?

**Comment by khajavi** [comment]
How does that logic resolve [this issue](https://github.com/zio/zio-http/issues/2776)? Will it handle the ctrl+c signal and release the acquired port?

**Comment by varshith257** [comment]
Yes, it will check for that porr if is already occupied it checks for another free port. So that there is no choice that Port bind error occurs

**Comment by khajavi** [comment]
The reason the user can't bind to that port is that on the last run, the port was not successfully released. So, you need to release it instead of choosing another port.

**Comment by varshith257** [comment]
@khajavi Here we go! Checkout this PR, the solution releases port and no changes to any cmds

https://github.com/zio/zio-quickstarts/pull/57

**Comment by guizmaii** [comment]
@khajavi, I disagree with you here. We need to improve the current examples by introducing sbt-revolver. 

> as we need to update all the tutorials on zio.dev referenced to this quickstart.

Well, that means that we need to update the doc for part of this work. 

But we should not introduce dedicated code to do this release like done here https://github.com/zio/zio-quickstarts/pull/57
This isn't a code example to give. That's not how things should be done.

**Comment by varshith257** [comment]
I am in the situation of two reviewers different perspectives 😅

Both have valid reasons. Yes, but if example semantics doesn't need to change as well as the newbie who starts zio with quickstarts can feel it overwhelmed with what's happening with this diff lines of  codes rather than being intended with perspective of showing him about Zio concepts. For this i agree with @guizmaii 

> To reduce the number of changes and make it consistent with other quick starts, I suggest resolving this issue without switching to alternative commands other than the sbt run.

For this i agree with @khajavi

**Comment by varshith257** [comment]
How about adding sbt-revolver plugin to quickstarts and add the single line in docs of quickstarts under a note  to redirect here for port binding errors
https://zio.dev/zio-http/installation/#hot-reload-changes-watch-mode

**Comment by khajavi** [comment]
@guizmaii  Overall, I understand and acknowledge your comments. Let me elaborate:

> @khajavi, I disagree with you here. We need to improve the current examples by introducing sbt-revolver.
> 
> > as we need to update all the tutorials on zio.dev referenced to this quickstart.
> 
> Well, that means that updating the doc for part of this work.

That was the concern about maintaining uniform material across documents and quickstart source codes, but not enforcing how to do that. As I mentioned, I'm not opposed to the `sbt-resolver` approach.

> But we should not introduce dedicated code to do this release like done here #57
Makes sense and I agree

> Yes, it does solve the issue because sbt-revolver solves the issue by introducing the `reStop` command which stops the server.

I mentioned that it's not fixed because the [reported issue](https://github.com/zio/zio-http/issues/2776) is expected when using `sbt run`. To resolve this issue, forking the JVM on the run command as a simple solution or adding `sbt-resolver` requires updating the corresponding quickstarts and tutorials on zio.dev as part of this effort; both are acceptable IMHO.

**Comment by varshith257** [comment]
@guizmaii What do you think ? Lets go with `sbt-revolver` plugin ?

**Comment by guizmaii** [comment]
@varshith257 Yes and part of your work here is to update the documentation too

**Comment by guizmaii** [comment]
@khajavi Thanks for your detailed answer :)

**Pull request closed without merging**